### PR TITLE
Client to work with ssl verification modes

### PR DESCRIPTION
**Pull request by mbenita-Cyberark**
### Desired Outcome

Client will be initialized with less parameters and specific, with a SslVerificationMetaData struct that holds the sslVerification data. This struct should be drilled down to the http_wrapper.

### Implemented Changes

- Added SslVerificationMetaData struct
- Changed constructor in both client and api
- Added functionality to create client from conjurrc (instead of the old way which was done inside the class constructor)

### Connected Issue/Story

Resolves [ONYX-14370](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-14370)

### Definition of Done
* Client could accept ssl_verification_meta_data object and all functionality will work

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: conjur/api/api.py [hunk 1]
```diff
@@ -48,32 +50,37 @@ class Api:
...
-            http_debug: bool = False,
-            login_id: str = None,
-            ssl_verify: bool = True,
-            url: str = None
+            url: str,
+            account: str,
+            credentials_provider: CredentialsStoreInterface,
+            ssl_verification_meta_data: SslVerificationMetaData = None,
+            debug: bool = False,
+            http_debug=False,
```
Comment (abrahamko) [inline conjur/api/api.py:58]:
```suggestion
            http_debug = False,
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,7 +14,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - The `hostfactory` method `create token` is now available in CLI and SDK to create a hostfactory token 
   to manage hosts and permissions in a dynamic way
   [cyberark/conjur-api-python3#339](https://github.com/cyberark/conjur-api-python3/pull/339)
-
+- Stop supporting `Client` initialization from disk.
```
Comment (mbenita-Cyberark) [inline CHANGELOG.md:17]: @shulifink @InbalZilberman

File: README.md [hunk 3]
```diff
@@ -96,11 +103,10 @@ API on it:
...
 
-client = Client(url='https://conjur.myorg.com',
-                account='default',
-                login_id='admin',
-                password='mypassword',
-                ca_bundle='/path/to/my/ca/bundle.pem')
+client = Client(conjurrc_data=ConjurrcData(...),
+                ssl_verification_mode=SslVerificationMode.WITH_TRUST_STORE.,
+                credentials_provider=FileCredentialsProvider(),
+                debug=False)
```
Comment (mbenita-Cyberark) [inline README.md:109]: @InbalZilberman

File: README.md [hunk 4]
```diff
@@ -111,26 +117,14 @@ new_value = client.get('conjur/my/variable')
...
-Write the code as in the first example but initialize the Client with the following arguments:
-
-```python3
-client = Client(url='https://conjur.myorg.com',
-                account='default',
-                login_id='admin',
-                api_key='myapikey',
-                ca_bundle='/path/to/my/ca/bundle.pem')
-```
-
```
Comment (mbenita-Cyberark) [inline README.md:125]: @InbalZilberman

File: README.md [hunk 5]
```diff
@@ -165,60 +159,48 @@ login admin
...
 ```
 
-If you store connection details in the `conjurrc` and use a credential store or `netrc` for 
-storing credentials, you can create the Client in the following way and the details will 
-be extracted implicitly:
-
-```
-client = Client()
-```
-
```
Comment (mbenita-Cyberark) [inline README.md:175]: @InbalZilberman

File: CHANGELOG.md (hunk 2 above)
Comment (InbalZilberman) [reply CHANGELOG.md:17]: LGTM

**Review by InbalZilberman — COMMENTED** [review]
LGTM

File: conjur/util/util_functions.py [hunk 6]
```diff
@@ -101,3 +103,20 @@ def get_current_os() -> OSTypes:  # pragma: no cover
     if platform.system() == "Windows":
         return OSTypes.WINDOWS
     return OSTypes.UNKNOWN
+
+
+def get_ssl_verification_meta_data_from_conjurrc(
```
Comment (abrahamko) [inline conjur/util/util_functions.py:108]:
```suggestion
def get_ssl_verification_metadata_from_conjurrc(
```

File: conjur/util/util_functions.py [hunk 7]
```diff
@@ -101,3 +103,20 @@ def get_current_os() -> OSTypes:  # pragma: no cover
...
+    """
+    if not conjur_data:
+        conjur_data = ConjurrcData.load_from_file()
+    cert_path = conjur_data.cert_file
+    if not ssl_verify:
+        return SslVerificationMetadata(SslVerificationMode.NO_SSL)
+    if not cert_path:
+        return SslVerificationMetadata(SslVerificationMode.WITH_TRUST_STORE)
+    if cert_path and cert_path != DEFAULT_CERTIFICATE_FILE:
+        return SslVerificationMetadata(SslVerificationMode.WITH_CA_BUNDLE, cert_path)
```
Comment (abrahamko) [inline conjur/util/util_functions.py:121]: Please add newline after each `if`

File: test/test_unit_api.py [hunk 8]
```diff
@@ -42,9 +47,19 @@
...
 MOCK_HOSTFACTORY_WITHOUT_CIDR_OBJECT = CreateTokenData(host_factory="some_host_factory", days=1)
+MockCredentials = CredentialsData(login='myuser', password='mypass')
+
+
+def create_api(url='http://localhost', account='default',
+               ssl_verification_mode=SslVerificationMode.SELF_SIGN,
+               cert_path=DEFAULT_CERTIFICATE_FILE):
+    return Api(conjurrc_data=ConjurrcData(conjur_url=url, account=account, cert_file=cert_path),
+               ssl_verification_mode=ssl_verification_mode,
+               credentials_provider=FileCredentialsProvider())
```
Comment (abrahamko) [inline test/test_unit_api.py:61]: If newline is needed between params, either in signature or method call, separate *each* param in newline.
Also, I prefer not calling methods within methods.
```suggestion
def create_api(url='http://localhost',
               account='default',
               ssl_verification_mode=SslVerificationMode.SELF_SIGN,
               cert_path=DEFAULT_CERTIFICATE_FILE):
    conjurrc_data = ConjurrcData(conjur_url=url, account=account, cert_file=cert_path)
    return Api(conjurrc_data=conjurrc_data,
               ssl_verification_mode=ssl_verification_mode,
               credentials_provider=FileCredentialsProvider())
```

File: test/test_unit_utils.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import unittest
+
+from conjur.api.models import SslVerificationMode
+from conjur.constants import DEFAULT_CERTIFICATE_FILE
+from conjur.util import util_functions as utils
+from conjur.data_object import ConjurrcData
+
+
+class UserInputDataTest(unittest.TestCase):
```
Comment (abrahamko) [inline test/test_unit_utils.py:9]: Good tests! simple and to the point.
You could combine them into single parametrized test, but it's as good as these, so it's only a hint.

**Review by abrahamko — DISMISSED** [review]

**Review by abrahamko — APPROVED** [review]
LGTM

**Review by abrahamko — APPROVED** [review]
LGTM

**Pull request merged**